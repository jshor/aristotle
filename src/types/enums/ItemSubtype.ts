enum ItemSubtype {
  None = 'None',

  CustomCircuit = 'CustomCircuit',

  /* input nodes */
  Switch = 'Switch',
  PushButton = 'PushButton',
  Clock = 'Clock',

  /* output nodes */
  Lightbulb = 'Lightbulb',
  DigitDisplay = 'DigitDisplay',

  /* buffers */
  Buffer = 'Buffer',
  Not = 'Not',

  /* tri-state */
  Tristate = 'Tristate',

  /* logic gates */
  And = 'And',
  Or = 'Or',
  Nand = 'Nand',
  Nor = 'Nor',
  Xnor = 'Xnor',
  Xor = 'Xor'
}

export default ItemSubtype
