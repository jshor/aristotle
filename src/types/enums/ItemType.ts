enum ItemType {
  Buffer = 'Buffer',
  CircuitNode = 'CircuitNode',
  InputNode = 'InputNode',
  IntegratedCircuit = 'IntegratedCircuit',
  LogicGate = 'LogicGate',
  OutputNode = 'OutputNode'
}

export default ItemType
