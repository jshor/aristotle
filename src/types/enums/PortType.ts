enum PortType {
  Output = 0,
  Input
}

export default PortType
