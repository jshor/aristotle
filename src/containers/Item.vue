<template>
  <draggable
    :position="position"
    :zoom="store.zoomLevel"
    :style="{
      transform: `rotate(${90 * rotation}deg)`,
      outline: 'none',
      zIndex
    }"
    :snap-mode="type === ItemType.Freeport ? 'radius' : 'outer'"
    :snap-boundaries="store.snapBoundaries"
    :bounding-box="boundingBox"
    :force-dragging="forceDragging"
    :aria-label="name"
    :title="name"
    :tabindex="0"
    @drag-start="onDragStart"
    @drag="onDrag"
    @drag-end="onDragEnd"
    @focus="onFocus"
    @blur="onBlur"
    @keydown.esc="onEscapeKey"
    @mousedown.stop="onMouseDown"
    ref="selectable"
    data-test="selectable"
  >
    <properties
      v-if="isPropertiesEnabled"
      :tabindex="0"
      :properties="properties"
      :id="id"
      @update="store.setProperties"
      @focus="onPropertiesFocus"
      @blur="onBlur"
      aria-label="Properties dialog"
      data-test="properties"
    />
    <selectable
      :is-selected="isSelected"
      :flash="flash"
    >
      <freeport v-if="type === ItemType.Freeport" />
      <integrated-circuit
        v-else-if="type === ItemType.IntegratedCircuit"
        :port-list="portList"
        @dblclick="$emit('openIntegratedCircuit')"
      />
      <input-switch
        v-else-if="type === ItemType.InputNode"
        :value="ports[0]?.value"
        @toggle="value => store.setPortValue({ id: ports[0].id, value })"
      />
      <lightbulb
        v-else-if="type === ItemType.OutputNode"
        :value="ports[0]?.value"
        :type="subtype"
      />
      <logic-gate
        v-else
        :input-count="(properties.inputCount?.value as number)"
        :type="subtype"
      />
      <port-set>
        <template
          v-for="(ports, orientation) in portList"
          v-slot:[orientation]
        >
          <port-item
            v-for="port in ports"
            :tabindex="0"
            :store="rawStore"
            :id="port.id"
            :key="port.id"
            :type="port.type"
            :is-freeport="port.isFreeport"
            :position="port.position"
            :orientation="port.orientation + rotation"
            :connected-port-ids="port.connectedPortIds"
            :rotation="rotation"
            @keypress.esc="onEscapeKey"
            @focus="onPortFocus(port.id)"
            @blur="$event => onPortBlur(port.id, $event)"
            @deselect="focus"
            data-test="port-item"
          />
        </template>
      </port-set>
    </selectable>
  </draggable>
</template>

<script lang="ts">
import ResizeObserver from 'resize-observer-polyfill'
import { StoreDefinition } from 'pinia'
import { defineComponent, PropType, onMounted, ref, ComponentPublicInstance, computed } from 'vue'
import Draggable from '@/components/Draggable.vue'
import LogicGate from '@/components/item/elements/LogicGate.vue'
import InputSwitch from '@/components/item/elements/InputSwitch.vue'
import IntegratedCircuit from '@/components/item/elements/IntegratedCircuit.vue'
import Lightbulb from '@/components/item/elements/Lightbulb.vue'
import Freeport from '@/components/item/Freeport.vue'
import PortHandle from '@/components/PortHandle.vue'
import PortSet from '@/components/item/PortSet.vue'
import Properties from '@/components/item/Properties.vue'
import Selectable from '@/components/Selectable.vue'
import PortItem from './PortItem.vue'
import ItemType from '@/types/enums/ItemType'
import ItemSubtype from '@/types/enums/ItemSubtype'
import DocumentState from '@/store/DocumentState'

/**
 * This component represents a two-dimensional item in the editor.
 * This can take on the form of a logic gate, input switch, lightbulb, user label, and more.
 * Items also contain ports, which are the circles that wires attach to to provide or send signals.
 */
export default defineComponent({
  name: 'Item',
  components: {
    Draggable,
    LogicGate,
    Freeport,
    InputSwitch,
    IntegratedCircuit,
    Lightbulb,
    PortHandle,
    PortSet,
    PortItem,
    Properties,
    Selectable
  },
  props: {
    /**
     * Bounding box of the item.
     */
    boundingBox: {
      type: Object as PropType<BoundingBox>,
      default: () => ({
        left: 0,
        top: 0,
        right: 0,
        bottom: 0
      })
    },

    /**
     * The directional rotation of the item.
     *
     * @values 0 = Left, 1, 2, 3
     */
    rotation: {
      type: Number,
      default: 0
    },

    /**
     * CSS z-index value.
     */
    zIndex: {
      type: Number,
      default: 1
    },

    /**
     * Whether or not the item is selected.
     */
    isSelected: {
      type: Boolean,
      default: false
    },

    /**
     * High-level taxonomy of the item.
     */
    type: {
      type: String as PropType<ItemType>,
      required: true
    },

    /**
     * More specific taxnommy of the item.
     */
    subtype: {
      type: String,
      required: true
    },

    /**
     * x-y position of the item in the editor.
     */
    position: {
      type: Object as PropType<Point>,
      required: true
    },

    /**
     * Optional user-friendly name of the item.
     */
    name: {
      type: String,
      default: 'Node'
    },

    /**
     * List of port IDs associated with the item.
     */
    portIds: {
      type: Array as PropType<string[]>,
      default: () => []
    },

    /**
     * Optional editable properties.
     */
    properties: {
      type: Object as PropType<PropertySet>,
      default: () => ({})
    },

    /**
     * ID of the group the item is associated to (if any).
     */
    groupId: {
      type: String,
      default: null
    },

    /**
     * Item ID.
     */
    id: {
      type: String,
      required: true
    },

    /** Whether or not the item should show a flash once to the user. */
    flash: {
      type: Boolean,
      default: false
    },

    /**
     * Document store instance.
     */
    store: {
      type: Function as PropType<StoreDefinition<string, DocumentState>>,
      required: true
    }
  },
  setup (props) {
    const store = props.store()
    const selectable = ref<ComponentPublicInstance<HTMLElement>>()
    const forceDragging = ref(false)

    let hasFocus = false
    let isDragging = false

    /* list of all ports that belong to this item */
    const ports = computed(() => {
      return props
        .portIds
        .map(portId => store.ports[portId])
    })

    /* slot-to-port-list map for displaying port-item elements */
    const portList = computed(() => {
      const locations = ['left', 'top', 'right', 'bottom']

      return ports.value.reduce((map: Record<string, Port[]>, port: Port) => ({
        ...map,
        [locations[port.orientation]]: [
          ...map[locations[port.orientation]],
          port
        ]
      }), locations.reduce((m, type) => ({
        ...m, [type]: []
      }), {
        left: [],
        top: [],
        right: [],
        bottom: []
      }))
    })

    /* whether or not the properties trigger button should be visible */
    const isPropertiesEnabled = computed(() => {
      return props.isSelected && hasFocus && Object.keys(props.properties).length > 0
    })

    /**
     * Places the window focus on the item element.
     */
    function focus () {
      selectable.value?.$el.focus()
      hasFocus = true
    }

    /**
     * Item drag start event handler.
     * This will compute the snappable boundaries on the editor.
     */
    function onDragStart () {
      store.setSnapBoundaries(props.id)
    }

    /**
     * Item drag event handler.
     * This will cache the undo-able state if dragged for the first time.
     *
     * @param position - new position that item has moved to
     */
    function onDrag (position: Point) {
      if (!isDragging) {
        isDragging = true
        store.cacheState() // cache the undo-able state in case a drag completes
      }

      store.setSelectionPosition({ id: props.id, position })
    }

    /**
     * Item drag end event handler.
     * This will commit the cached undo-able state.
     */
    function onDragEnd () {
      if (isDragging) {
        store.commitCachedState()
        isDragging = false
      }
    }

    /**
     * Inverts the selection state the item.
     */
    function onMouseDown ($event: MouseEvent) {
      if (props.isSelected) {
        store.deselectItem(props.id)
      } else {
        store.selectItem(props.id, $event.ctrlKey)
      }

      focus()
    }

    /**
     * Escape keydown event handler.
     * This will blur the element if neither it nor any of its children are focused.
     *
     * @param $event
     */
    function onEscapeKey ($event: KeyboardEvent) {
      if (document.activeElement === selectable.value?.$el) {
        selectable.value?.$el.blur()
      } else {
        $event.preventDefault()
      }
    }

    /**
     * Item blur event handler.
     * Updates `hasFocus` to false if the selection has left the item and any of its children.
     */
    function onBlur ($event: FocusEvent) {
      if (!selectable.value?.$el.contains($event.relatedTarget as Node)) {
        hasFocus = false
      }
    }

    /**
     * Item focus event handler.
     */
    function onFocus () {
      setTimeout(() => {
        if (!props.isSelected) {
          store.selectItem(props.id)
        }
      })

      hasFocus = true
    }

    /**
     * Properties dialog trigger button focus event handler.
     */
    function onPropertiesFocus () {
      store.setActivePortId(null)
      hasFocus = true
    }

    /**
     * Port focus event handler.
     *
     * @param portId - ID of the port focused
     */
    function onPortFocus (portId: string) {
      onFocus()
      store.setActivePortId(portId)
    }

    /**
     * Port blur event handler.
     *
     * @param portId - ID of the port blurred
     * @param $event - focus event
     */
    function onPortBlur (portId: string, $event: FocusEvent) {
      if (store.activePortId === portId) {
        store.setActivePortId(null)
      }
      onBlur($event)
    }

    /**
     * Item resize event handler. This will inform the store of the new size of the item.
     *
     * @param targets - target element that has been resized
     */
    function onSizeChanged ([ target ]: ResizeObserverEntry[]) {
      store.setItemSize({
        id: props.id,
        rect: target.contentRect
      })
    }

    const observer = new ResizeObserver(onSizeChanged)

    onMounted(() => {
      if (selectable.value) {
        observer.observe(selectable.value.$el)
      }

      if (store.activeFreeportId === props.id) {
        // if the item is an active freeport, then it is actively being drawn
        // allow for it to be moved when the mouse moves on first creation
        forceDragging.value = true
        store.activeFreeportId = null
        isDragging = true
      }

      store.setProperties({
        id: props.id,
        properties: props.properties
      })
    })

    return {
      selectable,
      store,
      ports,
      rawStore: props.store,
      portList,
      forceDragging,
      isPropertiesEnabled,
      focus,
      onDragStart,
      onDrag,
      onDragEnd,
      onMouseDown,
      onEscapeKey,
      onBlur,
      onFocus,
      onPropertiesFocus,
      onPortFocus,
      onPortBlur,
      onSizeChanged,
      ItemSubtype,
      ItemType
    }
  },
  watch: {
    isSelected (isSelected: boolean) {
      if (isSelected) {
        this.$el.focus()
      }
    }
  }
})
</script>
