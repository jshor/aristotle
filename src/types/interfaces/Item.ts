import ItemSubtype from '../enums/ItemSubtype'
import ItemType from '../enums/ItemType'
import ClockPulse from '@/store/document/oscillator/ClockPulse'
import Point from '@/types/interfaces/Point'
import BoundingBox from '../types/BoundingBox'
import BaseItem from './BaseItem'
import IntegratedCircuit from './IntegratedCircuit'
import Property from './Property'
import LogicValue from '../enums/LogicValue'
import ItemProperties from './ItemProperties'

interface Item extends BaseItem {
  type: ItemType
  subtype: ItemSubtype
  /** The default name to refer to this item. */
  defaultName: string
  /**
   * The current defined name to refer to this item.
   * This differs from {@link defaultName} in that it can be changed by the user.
   * It will be auto-generated by the editor when the item is first created.
   */
  name: string
  portIds: string[]
  groupId: string | null
  rotation: number
  boundingBox: BoundingBox
  position: Point
  zIndex: number
  width: number
  height: number
  properties: ItemProperties
  clock?: ClockPulse | null
  integratedCircuit?: IntegratedCircuit
}

export default Item
