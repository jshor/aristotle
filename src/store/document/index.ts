import { StoreGetters, defineStore } from 'pinia'
import { DocumentState, state } from './state'
import * as clipboardActions from './actions/clipboard'
import * as connectionActions from './actions/connection'
import * as editorActions from './actions/editor'
import * as groupingActions from './actions/grouping'
import * as itemActions from './actions/item'
import * as panningActions from './actions/panning'
import * as portActions from './actions/port'
import * as positioningActions from './actions/positioning'
import * as oscillatorActions from './actions/oscillator'
import * as rotationActions from './actions/rotation'
import * as selectionActions from './actions/selection'
import * as simulationActions from './actions/simulation'
import * as sizingActions from './actions/sizing'
import * as snappingActions from './actions/snapping'
import * as undoRedoActions from './actions/undoRedo'
import * as zIndexActions from './actions/zIndex'
import * as zoomingActions from './actions/zooming'
import * as getters from './getters'
import sortByZIndex from '@/utils/sortByZIndex'
import BaseItem from '@/types/interfaces/BaseItem'
import Item from '@/types/interfaces/Item'
import Connection from '@/types/interfaces/Connection'

export const createDocumentStore = (documentId: string) => defineStore({
  id: documentId,
  state,
  getters: {
    ...getters,

    baseItems (state) {
      return (Object
        .values(state.connections) as BaseItem[])
        .concat(Object.values(state.items) as BaseItem[])
        .sort(sortByZIndex)
    },

    hasSelection (state) {
      return state.selectedItemIds.size > 0 || state.selectedConnectionIds.size > 0
    },

    hasSelectedItems (state) {
      return state.selectedItemIds.size > 0
    },

    canGroup (state) {
      let ungroupedCount = 0
      let groupedCount = 0

      const hasUngrouped = (list: Set<string>, map: Record<string, Connection | Item>) => {
        for (const id of list) {
          map[id].groupId
            ? groupedCount++
            : ungroupedCount++

          if (ungroupedCount > 1 || (groupedCount && ungroupedCount)) {
            // at least two items are ungrouped, or at least one item is grouped and at least one item is ungrouped
            return true
          }
        }

        return false
      }

      return hasUngrouped(state.selectedItemIds, state.items) || hasUngrouped(state.selectedConnectionIds, state.connections)
    },

    canUngroup (state) {
      return state.selectedGroupIds.size > 0
    },

    canUndo (state) {
      return state.undoStack.length > 0
    },

    canRedo (state) {
      return state.redoStack.length > 0
    }
  },
  actions: {
    ...clipboardActions,
    ...connectionActions,
    ...editorActions,
    ...groupingActions,
    ...itemActions,
    ...panningActions,
    ...portActions,
    ...positioningActions,
    ...oscillatorActions,
    ...rotationActions,
    ...selectionActions,
    ...simulationActions,
    ...snappingActions,
    ...sizingActions,
    ...undoRedoActions,
    ...zIndexActions,
    ...zoomingActions
  }
})

export type DocumentStore = ReturnType<typeof createDocumentStore>

export type DocumentStoreInstance = ReturnType<DocumentStore>
