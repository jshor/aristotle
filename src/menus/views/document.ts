import { MenuFactory } from '@/types/interfaces/MenuFactory'
import { DocumentStore } from '@/store/document'
import { createHelpSubmenu } from '../submenus/help'
import { createEditSubmenu } from '../submenus/edit'
import { createFileSubmenu } from '../submenus/file'
import { createViewSubmenu } from '../submenus/view'
import { t } from '@/utils/i18n'

/**
 * Creates the document application menu.
 */
export const createDocumentMenu: MenuFactory = (useDocumentStore?: DocumentStore) => {
  return [
    {
      label: t('menu.file.parent'),
      submenu: createFileSubmenu(useDocumentStore)
    },
    {
      label: t('menu.edit.parent'),
      submenu: createEditSubmenu(useDocumentStore)
    },
    {
      label: t('menu.view.parent'),
      submenu: createViewSubmenu(useDocumentStore)
    },
    {
      label: t('menu.help.parent'),
      role: 'help',
      submenu: createHelpSubmenu()
    }
  ]
}
