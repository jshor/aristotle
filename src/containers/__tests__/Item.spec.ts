import { mount, VueWrapper } from '@vue/test-utils'
import { setActivePinia, createPinia } from 'pinia'
import { ComponentPublicInstance } from 'vue'
import Item from '../Item.vue'
import { createDocumentStore, DocumentStore, DocumentStoreInstance } from '@/store/document'
import { createItem, createPort, stubAll } from '@/store/document/actions/__tests__/__helpers__'
import ItemType from '@/types/enums/ItemType'
import PortType from '@/types/enums/PortType'
import ItemSubtype from '@/types/enums/ItemSubtype'
import { useRootStore } from '@/store/root'
import type ItemInterface from '@/types/interfaces/Item'
import Port from '@/types/interfaces/Port'

setActivePinia(createPinia())

describe('Item container', () => {
  let wrapper: VueWrapper
  let storeDefinition: DocumentStore
  let store: DocumentStoreInstance
  let item1: ItemInterface
  let port1: Port, port2: Port

  const props = {
    zIndex: 1,
    isSelected: false,
    id: 'item1',
    flash: false
  }

  beforeEach(() => {
    storeDefinition = createDocumentStore('test-id')
    store = storeDefinition()

    item1 = createItem('item1', ItemType.InputNode, { portIds: ['port1', 'port2'] })
    port1 = createPort('port1', item1.id, PortType.Input, { isMonitored: true, hue: 12 })
    port2 = createPort('port2', item1.id, PortType.Output)

    store.ports = { port1, port2 }
    store.items = { item1 }
    store.setItemPosition({ id: item1.id, position: { x: 0, y: 0 } })

    wrapper = mount(Item, {
      props: {
        ...props,
        store: storeDefinition
      }
    })
  })

  afterEach(() => jest.resetAllMocks())

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('when the item rotates', () => {
    it('should match the new rotation', () => {
      store.items.item1.rotation = 2

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when the item has variable input ports that have changed', () => {
    it('should match the new rotation', async () => {
      stubAll(store, ['setItemSize'])

      store.items.item1.properties = {
        inputCount: {
          value: 2,
          type: 'number',
          label: 'Input Count'
        }
      }

      await wrapper.vm.$nextTick()

      expect(store.setItemSize).toHaveBeenCalledWith({
        id: 'item1',
        rect: expect.any(Object)
      })
    })
  })

  describe('the properties button', () => {
    it('should be visible on a selected item', async () => {
      store.selectedItemIds = new Set(['item1'])

      await wrapper.setProps({ isSelected: true })

      expect(wrapper.findComponent({ name: 'Properties' }).exists()).toBe(true)
    })

    it('should not be visible on a non-selected item', async () => {
      await wrapper.setProps({ isSelected: false })

      expect(wrapper.findComponent({ name: 'Properties' }).exists()).toBe(false)
    })

    it('should not be visible when multiple items are selected', async () => {
      store.selectedItemIds = new Set(['item1', 'item2'])
      await wrapper.setProps({ isSelected: true })

      expect(wrapper.findComponent({ name: 'Properties' }).exists()).toBe(false)
    })
  })

  describe('when the item is double-clicked', () => {
    it('should reveal the custom circuit, if it exists', () => {
      const rootStore = useRootStore()

      stubAll(rootStore, ['openIntegratedCircuit'])

      item1.subtype = ItemSubtype.CustomCircuit
      wrapper.trigger('dblclick')

      expect(rootStore.openIntegratedCircuit).toHaveBeenCalled()
    })

    it('should not open any integrated circuits if the item is not one', () => {
      const rootStore = useRootStore()

      stubAll(rootStore, ['openIntegratedCircuit'])

      wrapper.trigger('dblclick')

      expect(rootStore.openIntegratedCircuit).not.toHaveBeenCalled()
    })
  })

  describe('when the item has begun to be dragged', () => {
    beforeEach(async () => {
      stubAll(store, [
        'cacheState',
        'setSnapBoundaries'
      ])

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('dragStart')
    })

    it('should cache the state', () => {
      expect(store.cacheState).toHaveBeenCalledTimes(1)
    })

    it('should compute the new snap boundaries', () => {
      expect(store.setSnapBoundaries).toHaveBeenCalledTimes(1)
    })
  })

  describe('when the item has is being dragged', () => {
    it('should set the new position of the state with respect to the zoom and its offset', async () => {
      const position = {
        x: 10,
        y: 20
      }
      const offset = {
        x: 5,
        y: 10
      }

      stubAll(store, ['dragItem'])

      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('drag', position, offset)

      expect(store.dragItem).toHaveBeenCalledTimes(1)
      expect(store.dragItem).toHaveBeenCalledWith('item1', {
        x: position.x - offset.x - store.items.item1.width / 2 * store.zoomLevel,
        y: position.y - offset.y - store.items.item1.height / 2 * store.zoomLevel
      })
    })
  })

  it('should blur the element if it is focused', async () => {
    const spy = jest.spyOn(wrapper.vm.$el, 'blur')

    jest
      .spyOn(document, 'activeElement', 'get')
      .mockReturnValue(wrapper.vm.$el)

    await wrapper
      .find('.port-handle')
      .trigger('keydown.esc')

    await new Promise(r => setTimeout(r))

    expect(spy).toHaveBeenCalled()
  })

  it('should show the hue of the port if it is monitored', async () => {
    const portHandle = await wrapper.findComponent({ name: 'PortHandle' })

    expect(portHandle.props().hue).toEqual(port1.hue)
  })

  it('should save preview a connection from a port when the space bar is hit', async () => {
    stubAll(store, ['cycleConnectionPreviews'])

    await wrapper
      .find('.port-handle')
      .trigger('keydown.space')

    expect(store.cycleConnectionPreviews).toHaveBeenCalledTimes(1)
    expect(store.cycleConnectionPreviews).toHaveBeenCalledWith('port1')
  })

  it('should save the previewed connection generated on a port when the enter key is hit', async () => {
    stubAll(store, ['commitPreviewedConnection'])

    await wrapper
      .find('.port-handle')
      .trigger('keydown.enter')

    expect(store.commitPreviewedConnection).toHaveBeenCalledTimes(1)
  })

  it('should save the previewed connection generated on a port when the enter key is hit', async () => {
    stubAll(store, ['commitPreviewedConnection'])

    await wrapper
      .find('.port-handle')
      .trigger('keydown.enter')

    expect(store.commitPreviewedConnection).toHaveBeenCalledTimes(1)
  })

  it('should set the active port ID when its context menu is shown', async () => {
    stubAll(store, ['setActivePortId'])

    await wrapper
      .find('.port-handle')
      .trigger('contextmenu')

    expect(store.setActivePortId).toHaveBeenCalledTimes(1)
    expect(store.setActivePortId).toHaveBeenCalledWith('port1')
  })

  it('should set the active port ID when a port is focused', async () => {
    stubAll(store, ['setActivePortId'])

    await wrapper
      .find('.port-handle')
      .trigger('focus')

    expect(store.setActivePortId).toHaveBeenCalledTimes(1)
    expect(store.setActivePortId).toHaveBeenCalledWith('port1')
  })

  it('should clear the active port ID when a port is blurred', async () => {
    stubAll(store, ['unsetActivePortId'])

    store.activePortId = 'port1'

    await wrapper
      .find('.port-handle')
      .trigger('blur')

    expect(store.unsetActivePortId).toHaveBeenCalledTimes(1)
    expect(store.unsetActivePortId).toHaveBeenCalledWith('port1')
  })

  describe('when the item is right-clicked', () => {
    const $event = new MouseEvent('contextmenu')

    beforeEach(() => stubAll(store, ['setItemSelectionState']))

    it('should select the item if not already selected', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('contextmenu', $event)

      expect(store.setItemSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setItemSelectionState).toHaveBeenCalledWith('item1', true)
    })

    it('should not select the item if it is already selected', async () => {
      await wrapper.setProps({ isSelected: true })
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('contextmenu', $event)

      expect(store.setItemSelectionState).not.toHaveBeenCalled()
    })

    it('should emit `contextmenu`', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('contextmenu', $event)

      expect(wrapper.emitted()).toHaveProperty('contextmenu')
      expect(wrapper.emitted().contextmenu[0]).toEqual([$event])
    })
  })

  describe('when the draggable element emits a selection gesture', () => {
    beforeEach(() => {
      stubAll(store, [
        'setItemSelectionState',
        'deselectAll'
      ])
    })

    it('should select the item', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('select')

      expect(store.setItemSelectionState).toHaveBeenCalledTimes(1)
      expect(store.setItemSelectionState).toHaveBeenCalledWith('item1', true)
    })

    it('should deselect other elements when the draggable component indicated to do so', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('select', true)

      expect(store.deselectAll).toHaveBeenCalledTimes(1)
    })

    it('should not deselect other elements when the draggable component did not indicate to do so', async () => {
      await wrapper
        .findComponent({ name: 'Draggable' })
        .vm
        .$emit('select')

      expect(store.deselectAll).not.toHaveBeenCalled()
    })
  })
})
