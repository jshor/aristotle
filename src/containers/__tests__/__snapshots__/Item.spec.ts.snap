// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item container should match the snapshot 1`] = `
"<div class="draggable" style="left: 0px; top: 0px; outline: none; pointer-events: none; z-index: 1;" data-test="item">
  <!--v-if-->
  <div class="circuit-component" style="transform: rotate(0deg);">
    <div class="input-node" type="InputNode" subtype="None" isselected="false" flash="false" name="" properties="[object Object]"><button type="button" class="input-node__flipper input-node__flipper--high-z" tabindex="0"><span>?</span></button></div>
    <div class="circuit-component__ports--left circuit-component__ports">
      <div class="port-pivot" style="transform: rotate(-0deg);" allow-touch-drag="" tabindex="0" title="" data-test="port-item">
        <div class="draggable" style="left: 0px; top: 0px;" tabindex="-1">
          <div class="port-handle" type="1" touch-friendly="">
            <div class="port-handle__display">
              <div class="port-handle__hue"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="port-pivot" style="transform: rotate(-0deg);" allow-touch-drag="" tabindex="0" title="" data-test="port-item">
        <div class="draggable" style="left: 0px; top: 0px;" tabindex="-1">
          <div class="port-handle" type="0" touch-friendly="">
            <div class="port-handle__display">
              <div class="port-handle__hue"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="circuit-component__ports--top circuit-component__ports"></div>
    <div class="circuit-component__ports--right circuit-component__ports"></div>
    <div class="circuit-component__ports--bottom circuit-component__ports"></div>
  </div>
</div>"
`;

exports[`Item container when the item rotates should match the new rotation 1`] = `
"<div class="draggable" style="left: 0px; top: 0px; outline: none; pointer-events: none; z-index: 1;" data-test="item">
  <!--v-if-->
  <div class="circuit-component" style="transform: rotate(0deg);">
    <div class="input-node" type="InputNode" subtype="None" isselected="false" flash="false" name="" properties="[object Object]"><button type="button" class="input-node__flipper input-node__flipper--high-z" tabindex="0"><span>?</span></button></div>
    <div class="circuit-component__ports--left circuit-component__ports">
      <div class="port-pivot" style="transform: rotate(-0deg);" allow-touch-drag="" tabindex="0" title="" data-test="port-item">
        <div class="draggable" style="left: 0px; top: 0px;" tabindex="-1">
          <div class="port-handle" type="1" touch-friendly="">
            <div class="port-handle__display">
              <div class="port-handle__hue"></div>
            </div>
          </div>
        </div>
      </div>
      <div class="port-pivot" style="transform: rotate(-0deg);" allow-touch-drag="" tabindex="0" title="" data-test="port-item">
        <div class="draggable" style="left: 0px; top: 0px;" tabindex="-1">
          <div class="port-handle" type="0" touch-friendly="">
            <div class="port-handle__display">
              <div class="port-handle__hue"></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="circuit-component__ports--top circuit-component__ports"></div>
    <div class="circuit-component__ports--right circuit-component__ports"></div>
    <div class="circuit-component__ports--bottom circuit-component__ports"></div>
  </div>
</div>"
`;
